edx/edx-enterprise
# [ENT-500] Add Program-based consent architecture

**Description:**
This pull request adds the ability to retrieve and post consent information scoped for an entire program at once.

This pull request also allows for treatment of course keys as first-class citizens in the consent process in addition to the existing use of course run keys; if a record with a course key is stored, it will answer to requests for any course run that's an instance of that course.

The programs consent feature takes advantage of this in that programs don't necessarily have canonical course runs; therefore, rather than consenting to any particular course run, the user is, behind the scenes, consenting to all the courses in the program.

**JIRA:** [ENT-500](https://openedx.atlassian.net/browse/ENT-500)

**Dependencies:** None

**Merge deadline:** 28 August 2017

**Installation requirements:**
- Course discovery set up
  - Program with courses
  - Catalog that includes all courses in the program
- EnterpriseCustomer set up
  - Catalog set to aforementioned catalog
  - Data sharing consent enabled/set to enforce at enrollment

**Testing instructions (API):**

1. Browse to URL matching pattern:
    ```
    /consent/api/v1/data_sharing_consent?enterprise_customer_uuid=your-ec-uuid-here&username=username&program_uuid=your-program-uuid-here
    ```
2. In the LMS shell, instantiate an API client with a staff user, aimed at `/consent/api/v1`.
3. Test the following with your client, verifying the expected behavior, and noting that calling the course-specific API with both course run IDs and the overall course IDs yields the expected result.:
    ```python
    # Start by verifying that no records appear to exist
    >>> client.data_sharing_consent.get(enterprise_customer_uuid='ec-uuid', username='username', program_uuid='program_uuid')
    {'enterprise_customer_uuid': 'ec-uuid', 'username': 'username', 'program_uuid': 'program_uuid', 'exists': False, 'consent_provided': False, 'consent_required': True}

    >>> client.data_sharing_consent.get(enterprise_customer_uuid='ec-uuid', username='username', course_id='course-v1:org+course+specific_course_run')
    {'enterprise_customer_uuid': 'ec-uuid', 'username': 'username', 'course_id': 'course-v1:org+course+specific_course_run', 'exists': False, 'consent_provided': False, 'consent_required': True}

    >>> client.data_sharing_consent.get(enterprise_customer_uuid='ec-uuid', username='username', course_id='org+course')
    {'enterprise_customer_uuid': 'ec-uuid', 'username': 'username', 'course_id': 'org+course', 'exists': False, 'consent_provided': False, 'consent_required': True}

    # Provide program-scoped consent
    >>> client.data_sharing_consent.post({'enterprise_customer_uuid': 'ec-uuid', 'username': 'username', 'program_uuid': 'program_uuid'})
    {'enterprise_customer_uuid': 'ec-uuid', 'username': 'username', 'program_uuid': 'program_uuid', 'exists': True, 'consent_provided': True, 'consent_required': False}

    # Verify that individual course run consent now has been provided
    >>> client.data_sharing_consent.get(enterprise_customer_uuid='ec-uuid', username='username', course_id='course-v1:org+course+specific_course_run')
    {'enterprise_customer_uuid': 'ec-uuid', 'username': 'username', 'course_id': 'course-v1:org+course+specific_course_run', 'exists': True, 'consent_provided': True, 'consent_required': False}

    # As well as for the course
    >>> client.data_sharing_consent.get(enterprise_customer_uuid='ec-uuid', username='username', course_id='org+course')
    {'enterprise_customer_uuid': 'ec-uuid', 'username': 'username', 'course_id': 'org+course', 'exists': True, 'consent_provided': True, 'consent_required': False}

    # Delete consent for all the items in the program
    >>> client.data_sharing_consent.delete(enterprise_customer_uuid='ec-uuid', username='username', program_uuid='program_uuid')
    {'enterprise_customer_uuid': 'ec-uuid', 'username': 'username', 'program_uuid': 'program_uuid', 'exists': True, 'consent_provided': False, 'consent_required': True}

    # Validate revocation on the course/course run
    >>> client.data_sharing_consent.get(enterprise_customer_uuid='ec-uuid', username='username', course_id='course-v1:org+course+specific_course_run')
    {'enterprise_customer_uuid': 'ec-uuid', 'username': 'username', 'course_id': 'course-v1:org+course+specific_course_run', 'exists': True, 'consent_provided': False, 'consent_required': True}

    >>> client.data_sharing_consent.get(enterprise_customer_uuid='ec-uuid', username='username', course_id='org+course')
    {'enterprise_customer_uuid': 'ec-uuid', 'username': 'username', 'course_id': 'org+course', 'exists': True, 'consent_provided': False, 'consent_required': True}
    ```

**Testing instructions (UI):**

1. Delete any consent records created in the API testing.
2. Create a new (non-staff) user account.
3. Log into that user.
4. Browse to the URL matching the following pattern, replacing the relevant UUIDs with your own (copy it to an external note; you'll need it again):
    ```
    http://localhost:18000/enterprise/grant_data_sharing_permissions?enterprise_customer_uuid=8019d1dc-1116-4e14-8629-2e5306c13c88&program_uuid=0594a613-08e4-45e7-92ca-39b488e46f34&next=https%3A%2F%2Fgoogle.com%2F&&failure_url=https%3A%2F%2Fnews.ycombinator.com%2F
    ```
5. Note that you see a consent prompt.
6. Decline consent.
7. Note that you're redirected to HackerNews (required parameter in the URL, and we don't have other handlers yet).
8. Browse to the same URL again.
9. This time, provide consent.
10. Note that you're redirected to Google (as above).
11. Attempt to browse to the URL again and note that you get a 404 because consent is no longer needed.
12. Delete the resulting consent record(s) from the database.
13. Repeat steps 4 through 10, adding an extra url parameter: `&enrollment_deferred=True`.
14. Repeat step 11, but note that the page is loadable again; the `enrollment_deferred` parameter resulted in the record not being committed to the database, which means it's still required.

**Merge checklist:**

  **DO NOT** just add dependencies to `requirements/*.txt` files.

**Post merge:**
